use_case_guide/real_time_dashboards.rst

```diff
@@ -138,141 +144,90 @@ and Citus will place matching shards on the same worker. This is called
-      -- never simultaneously write to the same shard.
-      IF pg_try_advisory_xact_lock(29999, split_part(p_dest_shard, '_', 4)::int) = false THEN
-        -- N.B. make sure the int constant (29999) you use here is unique within your system
-        RETURN;
-      END IF;
-
-      EXECUTE format($insert$
-        INSERT INTO %2$I (
-          site_id, ingest_time, request_count,
-          success_count, error_count, average_response_time_msec
-        ) SELECT
-          site_id,
-          date_trunc('minute', ingest_time) as minute,
-          COUNT(1) as request_count,
-          SUM(CASE WHEN (status_code between 200 and 299) THEN 1 ELSE 0 END) as success_count,
-          SUM(CASE WHEN (status_code between 200 and 299) THEN 0 ELSE 1 END) as error_count,
-          SUM(response_time_msec) / COUNT(1) AS average_response_time_msec
-        FROM %1$I
-        WHERE
-          date_trunc('minute', ingest_time)
-            > (SELECT COALESCE(max(ingest_time), timestamp '10-10-1901') FROM %2$I)
-          AND date_trunc('minute', ingest_time) < date_trunc('minute', now())
-        GROUP BY site_id, minute
-        ORDER BY minute ASC;
-      $insert$, p_source_shard, p_dest_shard);
-    END;
-    $$ LANGUAGE 'plpgsql';
-
-Inside this function you can see the dashboard query from earlier. It's been wrapped in
-some machinery which writes the results into ``http_request_1min`` and allows passing in
-the name of the shards to read and write from. It also takes out an advisory lock, to
-ensure there aren't any concurrency bugs where the same rows are written multiple times.
-
-The machinery above which accepts the names of the shards to read and write is necessary
-because only the coordinator has the metadata required to know what the shard pairs are. It has
-its own function to figure that out:
-
-.. code-block:: plpgsql
-
-    -- this function is created on the coordinator
-    CREATE FUNCTION colocated_shard_placements(left_table REGCLASS, right_table REGCLASS)
-    RETURNS TABLE (left_shard TEXT, right_shard TEXT, nodename TEXT, nodeport INT4) AS $$
-      SELECT
-        a.logicalrelid::regclass||'_'||a.shardid,
-        b.logicalrelid::regclass||'_'||b.shardid,
-        nodename, nodeport
-      FROM pg_dist_shard a
-      JOIN pg_dist_shard b USING (shardminvalue)
-      JOIN pg_dist_placement p ON (a.shardid = p.shardid)
-      JOIN pg_dist_node n ON (p.groupid = n.groupid)
-      WHERE a.logicalrelid = left_table AND b.logicalrelid = right_table;
-    $$ LANGUAGE 'sql';
-
-Using that metadata, every minute it runs a script which calls ``rollup_1min`` once for
-each pair of shards:
-
-.. code-block:: bash
-
-   #!/usr/bin/env bash
-
-   QUERY=$(cat <<END
-     SELECT * FROM colocated_shard_placements(
-       'http_request'::regclass, 'http_request_1min'::regclass
-     );
-   END
-   )
-
-   COMMAND="psql -h \$2 -p \$3 -c \"SELECT rollup_1min('\$0', '\$1')\""
+  CREATE OR REPLACE FUNCTION rollup_http_request() RETURNS void AS $$
```

Comment:
Is there a reason to put this in a function? Previously the function simplified some work around computing co-located placements, and calling the aggregation script for each pair. Now the function adds complexity to what is a very simple query. We can just note that for *actual* production use, we should consider wrapping this in a function and putting it in cron. Thoughts?

Reply:
In `rollup_http_request` I'm setting a variable `max_rollup_time` in one step, and using the variable in an INSERT statement in another. I had trouble trying to do it all in a single query.